### ISSUE TITLE
Future plans, goals, testing, and ideas for ParmEd

### ISSUE THREAD

**Issue by swails** [ParmEd/ParmEd#147]
I'm definitely happy with the recent increase in interest in ParmEd as well as a more well-formed image of some of the goals I have for it.  So I thought I'd try to solicit input on some ideas and questions I had in hopes of improving the development, direction, and tooling of ParmEd for those that may use it in their work.

The main goals I have for ParmEd are different, but complementary:

- Provide a flexible object describing a molecule complete with an arbitrary force field that is easy to manipulate in a high-level way
- Provide a platform that can read and write force field descriptions of molecules in any number of formats from different programs to allow for interconversions between the two of them
- Provide functionality that is useful to people only using one package (e.g., a user wants to interrogate or manipulate a Gromacs topology in Python chooses to use ParmEd, even if they don't want to convert to Amber, for instance)

I'm certainly open to other developments as well, that's just my checklist.  Toward that end, ideas I have are:

- More parsers (e.g., LAMMPS and Desmond)
- Command-line program that acts as a shortcut to convert files from one format to another (either as part of `parmed.py` or another operating mode)
- Improved website and documentation, including docs for developers (so people that can't pick up my design as easily as @leeping did can get started a bit easier)

I would also like a more rigorous, automated test suite like what @leeping has been using to great effect here so we know the relative accuracies between the different programs for a wide array of systems.  These can't be run on Travis due to time and resource constraints, but it would be nice to have them.  Would they be best as a subdirectory in this repository?  Or would it be better to have them in another repo in the ParmEd organization?  If the development of the tests and ParmEd will evolve together, it's best to keep them together.  If they're independent, it might be better to keep them separate so if the tests grow too large for Github, we can just move it off without inflating the size of this repo.

As part of the mission of being a portal connecting all of the simulation packages out there, perhaps it would also be nice to provide a repository of input files for each of the programs that do different things.  Similar to what I have at http://github.com/swails/Mdins for Amber.  Only the goal would be for the input files to be better organized and annotated.  This could even be used as a place to put input files used for papers... I don't know.  There are lots of possibilities.

**Comment by hainm** [comment]
start thinking about doc for developer now? :P

**Comment by leeping** [comment]
Hi Jason,

This topic has been revisited from time to time, but what do you think about developing a simplified MD simulation tool?  We already had a long discussion at https://github.com/omnia-md/omnia-md/issues/13 .  I think we agreed that a single command line tool can't take care of all use cases, but one important use case was a Python based front-end to introduce OpenMM to non-programmers.

I think that for ParmEd to add a command line front-end for OpenMM, it's mainly a matter of simplification and presentation.  `parmed.py` might have too many simulation editing facilities for new users, plus the name sounds like an editor.  We could add a script with another name that prints something like "ParmEd and OpenMM inside".

- Lee-Ping

**Comment by leeping** [comment]
The validation tests with other programs is definitely helpful and reassuring, and helps with developing new topology / parameter I/O features for other programs.  One possibility is that the test program (e.g. `atest.py` and `nifty.py`) and files could live in a subdirectory, e.g. `examples/validation` with instructions for running them.  I'd be happy for code to live in the ParmEd directory tree rather than be a different repo under the same organization, just because it's easier to keep track of.  There are more repos than I can remember these days.

In my opinion, the tests need to be kept simple in order to be effective, i.e. providing energy components for inspection and taking energy and force differences.  If we keep it that way, it would be sustainable to call the other code for energy / forces and parse the output.  It would be harder to test for consistency in a MD simulation for example.

**Comment by swails** [comment]
> start thinking about doc for developer now? :P

Dev docs are the hardest ones to make in my experience.  The best docs out there are the docstrings and comments in the code itself (so long as they are kept up-to-date), although I may jot down a basic style guide or something like that (and really, that's all Amber has...)

User docs obviously take precedence

**Comment by swails** [comment]
> I think that for ParmEd to add a command line front-end for OpenMM, it's mainly a matter of simplification and presentation. parmed.py might have too many simulation editing facilities for new users, plus the name sounds like an editor. We could add a script with another name that prints something like "ParmEd and OpenMM inside".

That's a good idea, I had forgotten about that long discussion.  One of the features that I think might make ParmEd a nice source of machinery for this (rather than, say, the OpenMM Python application layer) is the automatic file type detection employed by `chemistry.load_file`, which should make it pretty flexible as far as using any file types.  I still think `xparmed.py` provides some useful GUI widgets that we could package into a GUI application for the real newbies.

We can always scratch out a few ideas in a rough prototype and see what other people think of them.

**Comment by swails** [comment]
> One possibility is that the test program (e.g. atest.py and nifty.py) and files could live in a subdirectory, e.g. examples/validation with instructions for running them. I'd be happy for code to live in the ParmEd directory tree rather than a different repo under the same organization, just because it's easier to keep track of. There are more repos than I can remember these days.

We'll see if more people weigh in, but I think dumping this stuff in a `validation` folder (I'd vote for top-level directory as opposed to an `examples` subdir) is a good move for now.  We can always split later (and preserve all of the history using `git-subtree`) if we find that we need to.  And you're right that it's better not to be buried in more git repos than we can manage.

> In my opinion, the tests need to be kept simple in order to be effective, i.e. providing energy components for inspection and taking energy and force differences. If we keep it that way, it would be sustainable to call the other code for energy / forces and parse the output. It would be harder to test for consistency in a MD simulation for example.

Agreed.  One idea that I think would be nice for doing MD validation would be to make a `ValidationReporter` class for an OpenMM simulation that would just check the energies and forces and keep a running tab of agreement with other packages during the course of a simulation. I've started work on an `AmberValidationReporter` based on the use of the sander API, but I'd have to dig that up.

This may also be something worth bringing the greater OMM community into and seeing if they want to get involved and publish an extensive cross-program validation suite.

**Comment by hainm** [comment]
>>start thinking about doc for developer now? :P

>Dev docs are the hardest ones to make in my experience. The best docs out there are the docstrings and comments in the code itself (so long as they are kept up-to-date), although I may jot down a basic style guide or something like that (and really, that's all Amber has...)

>User docs obviously take precedence

I think  you just need to write bullet points about
* how you want others to follow ParmEd's style
* things should be done: like using delayed importing, ...
* point out where dev should look at your code
...

A quick overview (with figures?) is great place to start since potential developers might have different programming skills. 

I myself prefer short doc in our code. Very long one just distract me. :D (and it's really great that you don't copy and paste very long copyright in each of ParmEd's files :D).

**Comment by kyleabeauchamp** [comment]
Isn't the style guide here basically the same as what we're doing in mdtraj, which is in turn the "sklearn / astropy" scientific python convention?

**Comment by hainm** [comment]
I think so. But might be more detail, like the section ` SciPy structure` in [scipy doc] (http://docs.scipy.org/doc/scipy/reference/hacking.html)

**Comment by kyleabeauchamp** [comment]
For now, I think we could get away with just linked to 3 or 4 other "best practices" guides:

1. MDTraj
2.  Scipy
3.  sklearn
4.  astropy

Then we just need to document the "extra" idioms that we use for testing, continuous integration, delayed imports `_import()`

**Comment by hainm** [comment]
Agree. as long as we explicitly point to the source.

**Comment by swails** [comment]
FTR, ParmEd doesn't do 'lazy' or 'delayed' importing.  Importing the entire package, which brings in all of its dependencies and everything else at import time, takes 0.49 seconds on my laptop.  Anybody worried about that kind of initialization time should not be using ParmEd (as it is *not* the fastest set of parsers out there).

**Comment by kyleabeauchamp** [comment]
I was talking about the wrapping the import of optional dependencies to avoid unnecessary exceptions and to skip specific tests.

**Comment by swails** [comment]
As for style, it's PEP-8 with `numpydoc` docstrings that get processed into API docs by Sphinx.  The general rule I follow when contributing to other peoples' code is to follow existing styles (with regards to line length, indentation, and variableNaming, ClassNaming, or module_naming conventions), which definitely works here (I've occasionally bent this rule in the past when porting existing code to another code base with respect to private methods that should never be called, but the public-facing functions always follow the projects' existing style).

(BTW, I know this thread does not constitute documentation, it's simply stating what they are as part of the discussion).

**Comment by hainm** [comment]
>I was talking about the wrapping the import of optional dependencies to avoid unnecessary exceptions and to skip specific tests.

@swails: yes, this is what I was talking about too "_import("numpy")"

**Comment by swails** [comment]
> I was talking about the wrapping the import of optional dependencies to avoid unnecessary exceptions and to skip specific tests.

True, I do do that.  But it's still not "delayed importing" in the sense that (I *think*) `mdtraj` uses.  If numpy is available, it's imported immediately, not just as soon as some function needs it.

**Comment by swails** [comment]
> @swails: yes, this is what I was talking about too "_import("numpy")"

Yea, I don't do that.  It's more

```Python
try:
    import numpy as np
except ImportError:
    np = None
```

I haven't found it worth implementing (or copying someone else's implementation), nor do i find it particularly "cleaner" than what I showed above, really, and it gets in the way of linters like pyflakes, I think.

**Comment by hainm** [comment]
ah, if ParmEd does not depend much on other packages (like `mdtraj`), you don't need to do it.

**Comment by leeping** [comment]
> > One possibility is that the test program (e.g. atest.py and nifty.py) and files could live in a subdirectory, e.g. examples/validation with instructions for running them. I'd be happy for code to live in the ParmEd directory tree rather than a different repo under the same organization, just because it's easier to keep track of. There are more repos than I can remember these days.

> We'll see if more people weigh in, but I think dumping this stuff in a validation folder (I'd vote for top-level directory as opposed to an examples subdir) is a good move for now. We can always split later (and preserve all of the history using git-subtree) if we find that we need to. And you're right that it's better not to be buried in more git repos than we can manage.

I'd be fine with a top level directory too. :)

> > In my opinion, the tests need to be kept simple in order to be effective, i.e. providing energy components for inspection and taking energy and force differences. If we keep it that way, it would be sustainable to call the other code for energy / forces and parse the output. It would be harder to test for consistency in a MD simulation for example.

> Agreed. One idea that I think would be nice for doing MD validation would be to make a ValidationReporter class for an OpenMM simulation that would just check the energies and forces and keep a running tab of agreement with other packages during the course of a simulation. I've started work on an AmberValidationReporter based on the use of the sander API, but I'd have to dig that up.

This is a great idea.  I have a `TinkerValidationReporter` that I used to debug the AMOEBA implementation in OpenMM with @peastman a couple years ago.  There were some insidious bugs that only showed up after a neighbor search update or a Monte Carlo barostat step.  I'd be happy to contribute it.

I think the next level of detail for validation reporters is to compare the energy components (I usually print them out and check visually).  Doing this automatically means we need to parse the energy components somewhat reliably and understand which energy terms correspond to one another in different codes.  Do you think it would be nice to have this, or not worth the effort?

**Comment by swails** [comment]
> This is a great idea. I have a TinkerValidationReporter that I used to debug the AMOEBA implementation in OpenMM with @peastman a couple years ago. There were some insidious bugs that only showed up after a neighbor search update or a Monte Carlo barostat step. I'd be happy to contribute it.

This is actually where I got the idea :) (you told me about that last summer).  I have a place where I'm putting reporters in ParmEd (`chemistry/openmm/reporters.py`), although maybe we would want a separate `validationreporters` module.  Ideally we would support the major programs out there (Gromacs, Amber, Tinker, etc.), but I'd like to wait on this until we have Gromacs support working more reliably.

> I think the next level of detail for validation reporters is to compare the energy components (I usually print them out and check visually). Doing this automatically means we need to parse the energy components somewhat reliably and understand which energy terms correspond to one another in different codes. Do you think it would be nice to have this, or not worth the effort?

Definitely.  This is part of the reason why I by default assign each force to a different group in ParmEd (see [here](https://github.com/ParmEd/ParmEd/blob/master/chemistry/structure.py#L256-L267)) and have a convenience function to extract each group's contribution (see [here](https://github.com/ParmEd/ParmEd/blob/master/chemistry/openmm/utils.py#L7-L49)).  It's just up to the reporter to figure out how to best compare the "other" program's decomposition with the one OpenMM makes available.  It would also help to compare forces, too, I think.
